### PR TITLE
Refactor quantifier macros preprocessing pass

### PR DESCRIPTION
**Pull request by yoni206**
Moved `src/theory/quantifiers/macros.{cpp,h}` to `src/preprocessing/passes/quantifier_macros.{cpp,h}`,
and added the necessary methods for inheritance from PreprocessingPass.

No need to add a test - regress0 fails when adding Assert(false) when assertions had changed.

### CODE REVIEW

**Comment by ajreynol** [comment]
I'd like to get this merged in since I'd like to work on the macros module further.

**Comment by ajreynol** [comment]
Also, remember that the changes from #2038 need to be incorporated into this.

**Comment by 4tXJ7f** [comment]
@yoni206 Can you resolve the conflicts? I'll review after that.

File: src/preprocessing/passes/quantifier_macros.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ **
+ ** \brief Pre-process step for detecting quantifier macro definitions
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef __CVC4__PREPROCESSING__PASSES__QUANTIFIER_MACROS_H
+#define __CVC4__PREPROCESSING__PASSES__QUANTIFIER_MACROS_H
+
+#include <iostream>
```
Comment (4tXJ7f) [inline src/preprocessing/passes/quantifier_macros.h]: This include does not seem to be required.

File: src/smt/smt_engine.cpp [hunk 2]
```diff
@@ -4203,13 +4208,9 @@ void SmtEnginePrivate::processAssertions() {
...
-      quantifiers::QuantifierMacros qm( d_smt.d_theoryEngine->getQuantifiersEngine() );
-      bool success;
-      do{
-        success = qm.simplify( d_assertions.ref(), true );
-      }while( success );
-      //finalize the definitions
-      qm.finalizeDefinitions();
+      d_preprocessingPassRegistry.getPass("quantifier-macros")
+          ->apply(&d_assertions);
+      Trace("smt") << "POST macrosQuant" << endl;
```
Comment (4tXJ7f) [inline src/smt/smt_engine.cpp]: Nit: remove `Trace`.

File: src/preprocessing/passes/quantifier_macros.cpp [hunk 3]
```diff
@@ -27,19 +27,36 @@
...
 
+namespace CVC4 {
+namespace preprocessing {
+namespace passes {
 
-QuantifierMacros::QuantifierMacros( QuantifiersEngine * qe ) : d_qe( qe ){
+QuantifierMacros::QuantifierMacros(PreprocessingPassContext* preprocContext)
+    : PreprocessingPass(preprocContext, "quantifier-macros")
+{
   d_ground_macros = false;
```
Comment (4tXJ7f) [inline src/preprocessing/passes/quantifier_macros.cpp]: Nit: can be moved to initialization list.

File: src/preprocessing/passes/quantifier_macros.cpp [hunk 4]
```diff
@@ -27,19 +27,36 @@
...
+PreprocessingPassResult QuantifierMacros::applyInternal(
+    AssertionPipeline* assertionsToPreprocess)
+{
+  bool success;
+  do
+  {
+    success = simplify(assertionsToPreprocess->ref(), true);
+  } while (success);
+  finalizeDefinitions();
+  return PreprocessingPassResult::NO_CONFLICT;
```
Comment (4tXJ7f) [inline src/preprocessing/passes/quantifier_macros.cpp]: We should probably clear all maps (and maybe `d_ground_macros` for consistency) before returning. In the incremental case, we probably don't want state from earlier calls to `Smt::Engine::processAssertions()`.

File: src/preprocessing/passes/quantifier_macros.h [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+{
+ public:
+  QuantifierMacros(PreprocessingPassContext* preprocContext);
+  ~QuantifierMacros() {}
+ protected:
+  PreprocessingPassResult applyInternal(
+      AssertionPipeline* assertionsToPreprocess) override;
+
+ private:
+  bool d_ground_macros;
```
Comment (4tXJ7f) [inline src/preprocessing/passes/quantifier_macros.h]: Nit: move to other data members?

**Pull request merged**